### DISCUSSION TITLE
Deploying inside Next.js API Route

### DISCUSSION THREAD

**Discussion by Industrial** [directus/directus#5309] [Feature Requests]
Hi!

I'm using Next.js and I'm using the Serverless Framework with the Serverless Next.js component for deploying it to the cloud (Amazon AWS CloudFront with Lambda@Edge). This leads to very snappy and fast server side rendered applications.

Next.js supports server side logic using API routes which sit on `/api`. For example as part of the authentication strategy I'm using NextAuth which has an API handler for OAuth verifications at `/api/[...nextauth].tsx`.

What I would like to do if possible is to mount Directus under an API route at `/api/directus.tsx`. This would make the deployment process very easy as it would be wrapped into the serverless bundle and served from Lambda@Edge. No other cloud machinery needed.

Is that possible? Has someone attempted to do this before?

gr,

Tom

**Comment by rijkvanzanten** [comment]
Hi @Industrial!


I've never tried it! Technically speaking, the API opens connections to the database/redis/email etc during startup which are kept alive (pooled) during the lifespan of the API. I'm assuming that instead of using `.listen` on a server, you'd need to use the express route directly for it to play nicely with Next's setup?

You can theoretically import the express app instance (that doesn't listen) through `directus/dist/app`, but it will probably fail to properly connect/destroy those connections to the external services


Groetjes!

**Reply by Industrial** [reply]
Hi!

This is how far I've got:

```
ACCESS_TOKEN_TTL="15m"
CACHE_ENABLED=false
DB_CLIENT="pg"
DB_DATABASE="development"
DB_HOST="postgresql"
DB_PASSWORD="develop"
DB_PORT="5432"
DB_SSL="false"
DB_USER="postgresql"
DIRECTUS_HOST=directus
DIRECTUS_PORT=3001
EMAIL_FROM="no-reply@directus.io"
EMAIL_SENDMAIL_NEW_LINE="unix"
EMAIL_SENDMAIL_PATH="/usr/sbin/sendmail"
EMAIL_TRANSPORT="sendmail"
EXTENSIONS_PATH="./src/pages/api/directus/extensions"
KEY="85834406-bb25-4d5a-9b85-c6851e016e81"
PORT=3001
PUBLIC_URL="/"
RATE_LIMITER_DURATION=1
RATE_LIMITER_ENABLED=false
RATE_LIMITER_POINTS=25
RATE_LIMITER_STORE=memory
REFRESH_TOKEN_COOKIE_SAME_SITE="lax"
REFRESH_TOKEN_COOKIE_SECURE=false
REFRESH_TOKEN_TTL="7d"
SECRET="MdcI08XP_43s2zlV5RV1B__Zvnjy91Me"
STORAGE_LOCAL_DRIVER="local"
STORAGE_LOCAL_ROOT="./uploads"
STORAGE_LOCATIONS="local" 
```

```js
// Next.js proxy rewrite rules.
async redirects() {
  return [
    {
      source: '/activity/:path*',
      destination: '/api/directus/activity/:path*',
      permanent: false,
    },
    {
      source: '/admin/:path*',
      destination: '/api/directus/admin/:path*',
      permanent: false,
    },
    {
      source: '/assets/:path*',
      destination: '/api/directus/assets/:path*',
      permanent: false,
    },
    {
      source: '/auth/:path*',
      destination: '/api/directus/auth/:path*',
      permanent: false,
    },
    {
      source: '/collections/:path*',
      destination: '/api/directus/collections/:path*',
      permanent: false,
    },
    {
      source: '/custom/:path*',
      destination: '/api/directus/custom/:path*',
      permanent: false,
    },
    {
      source: '/extensions/:path*',
      destination: '/api/directus/extensions/:path*',
      permanent: false,
    },
    {
      source: '/fields/:path*',
      destination: '/api/directus/fields/:path*',
      permanent: false,
    },
    {
      source: '/files/:path*',
      destination: '/api/directus/files/:path*',
      permanent: false,
    },
    {
      source: '/folders/:path*',
      destination: '/api/directus/folders/:path*',
      permanent: false,
    },
    {
      source: '/fonts/:path*',
      destination: '/api/directus/fonts/:path*',
      permanent: false,
    },
    {
      source: '/graphql/:path*',
      destination: '/api/directus/graphql/:path*',
      permanent: false,
    },
    {
      source: '/items/:path*',
      destination: '/api/directus/items/:path*',
      permanent: false,
    },
    {
      source: '/permissions/:path*',
      destination: '/api/directus/permissions/:path*',
      permanent: false,
    },
    {
      source: '/presets/:path*',
      destination: '/api/directus/presets/:path*',
      permanent: false,
    },
    {
      source: '/relations/:path*',
      destination: '/api/directus/relations/:path*',
      permanent: false,
    },
    {
      source: '/revisions/:path*',
      destination: '/api/directus/revisions/:path*',
      permanent: false,
    },
    {
      source: '/roles/:path*',
      destination: '/api/directus/roles/:path*',
      permanent: false,
    },
    {
      source: '/server/:path*',
      destination: '/api/directus/server/:path*',
      permanent: false,
    },
    {
      source: '/settings/:path*',
      destination: '/api/directus/settings/:path*',
      permanent: false,
    },
    {
      source: '/users/:path*',
      destination: '/api/directus/users/:path*',
      permanent: false,
    },
    {
      source: '/utils/:path*',
      destination: '/api/directus/utils/:path*',
      permanent: false,
    },
    {
      source: '/webhooks/:path*',
      destination: '/api/directus/webhooks/:path*',
      permanent: false,
    },
  ]
}
```

```ts
// pages/api/[...directus].ts

import { IncomingMessage, ServerResponse } from 'http'

import createDirectusApplication from 'directus/dist/app'

export default async (req: IncomingMessage, res: ServerResponse): Promise<void> => {
  const handler = await createDirectusApplication()

  // Somehow this was unset at this point so I hardcoded it.
  process.env.PUBLIC_URL = '/api/directus'

  // This is where I take out part of the URL again for directus to process it
  // properly. Maybe in the future there could be an ENV var like BASE_PATH to
  // allow directus to be mounted under a prefix.
  req.url = req.url?.replace('/api/directus', '')

  console.log(`${req.method} ${req.url}`)

  handler(req, res)
}
```

This setup mounts directus as an API route in Next.js so *may* be deployed in a serverless environment (but doesn't have to).

Due to the rewrites I got everything working correctly and all the requests served by Directus up until the point where it handles the `POST /login`. This handler isn't returning/ending the response so the request hangs.

I'm not sure where to take it from here, but I do think that getting Directus to work together with Next.js without writing a plugin/NPM module would prove to be beneficial to the Directus ecosystem as people could easily deploy it as one unit. A lot of people nowadays use a GraphQL backend (like apollo server) as an API route in Next.js. I would prefer to use Directus :)

**Reply by rijkvanzanten** [reply]
That's awesome! Do you have a public repo somewhere with this setup so far? I'd love to play with it myself some more (esp to check out why login hangs), but I'm not familiar with NextJS

**Reply by Industrial** [reply]
I ran into a very dumb problem. Somehow I had an old yarn.lock file when I tried to integrate Directus into my project and it was all working. When I created a new repository (with `yarn create next-app`) and integrated Directus into that project it didn't work.

The project is https://github.com/Industrial/next-with-directus

```
git clone https://github.com/Industrial/next-with-directus.git
cd next-with-directus
yarn
yarn dev
```

then open http://localhost:3000/api/directus

It will give an Internal Server Error and the log will be

```
TypeError: Cannot read property 'LOG_LEVEL' of undefined
    at Object.<anonymous> (/location/next-with-directus/node_modules/directus/dist/logger.js:8:44)
    at Module._compile (internal/modules/cjs/loader.js:1063:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14)
    at Module.require (internal/modules/cjs/loader.js:952:19)
    at require (internal/modules/cjs/helpers.js:88:18)
    at Object.<anonymous> (/location/next-with-directus/node_modules/directus/dist/env.js:14:34)
    at Module._compile (internal/modules/cjs/loader.js:1063:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
```

It might have something to do with webpack (webpack 5 is used by next.js) or babel? Looking at the source code for `dist/logger.js` it should work.

**Reply by rijkvanzanten** [reply]
Ehh I don't know! There's too many variables to accurately say what might be wrong from my point of view.. FWIW, we rely on NPM Workspaces to properly link and propagate the different packages in dev mode, so that might cause some mismatches with yarn, but other than that, I'm not sure!

**Comment by jaads** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
